### ISSUE TITLE
Bug? Set list2 to list1 makes no copy but makes list2 same object als list1

### ISSUE THREAD

**Issue by Costor** [jmoenig/Snap--Build-Your-Own-Blocks#1634]
Seems that a bug in list assignment has been introduced some hours ago:  
At least code that worked fine until yesterday does no longer work:

Setting list2 to list1 does not copy contents of list1 into list2, but list2 becomes alias of list1.  Everything done to list2 is done to list1. 

See example (Press green flag, press space or 'a' key):

http://snap.berkeley.edu/snapsource/snap.html#present:Username=costor&ProjectName=SetListBug

Or is it a new feature (i.e. object assignment of lists?), and we need to explicitly assign copies of list1?

Kind regards, Costor

**Comment by jmoenig** [comment]
It's Math, not a bug.

set POTUS to @realDonaldTrump assigns the actual guy to the presidency, not a copy of him. 

I wish Americans had known this. Really.

Von meinem iPhone gesendet

**Comment by dekrain** [comment]
![memory](https://cloud.githubusercontent.com/assets/19158291/22177118/5abd474c-e018-11e6-998e-7beba830aeda.png)

**Comment by Costor** [comment]
Sorry, I hadn't checked this thoroughly beforehand. I had realized that < all but the first of list > is copying instead of giving a reference to the rest of the list, and also there is no < copy list > primitive, so I had come to think that assignment would also be copying. Treating lists/arrays like atomic variables on asigment (i.e. copying) also seemed natural in a language for kids who have no notion of a pointer.

**Comment by brianharvey** [comment]
All but first of doesn't copy if the list is a linked list.  And if it's an array, the copy is linked, so that the _next_ all but first of doesn't have to copy.

The object of the exercise is that if, for a given list, you are consistent in using it functionally or imperatively, in at most one copy it becomes linked or array, respectively, and then everything happens in constant time.

**Comment by micahcowan** [comment]
So... personally, I find this very dismaying. Differing performance is one thing, but differing behavior, especially due to reasons that may not be under the author's control (e.g., a blocks-library author has no control over how the list passed to it was used prior to passing) is very surprising behavior for a language. At least when, in both cases, they look, taste and smell the same as far as any standard blocks are concerned. There's no (builtin) block for "is this a linked list or an array?", and if there were (or one used a non-builtin version), having to use it would feel very much like a kludge, and a weakness in the language.

Naturally, if your library function wants to behave as though its argument were a linked-list, a workaround would be to place an element "in front of" the passed list, and then immediately discard it with "all but first of", but having to do this _also_ feels like a kludge, and a weakness in the language.

Would it be acceptable to automatically translate the referred-to-list to a (partially) linked list, before performing any linked-list operations? I really feel that while performance can vary based on past usage, behavior really, really shouldn't. A "learning" language like this shouldn't have such a dark corner (IMO).

IIRC, in all cases, a reference to a list is, under the hood, a reference to a JS object that indicates its linked-ness or otherwise, and then has a ref to the "real" list, so this transformation would not impact the behavior of other referrers (except in the unifying way I'm describing). We certainly do the transformation the other way whenever it's treated as an array...

**Comment by micahcowan** [comment]
Specifically, in a language like Java/C#/C++, etc, one fully expects `List<Foo>` and `Foo[]` to have potentially very different behaviors (though unfortunately none of those languages actually let you get a `all but first of ~` from `List<>` AFAIK). But in our case, we call both of these `list`, and make it non-trivial to distinguish, pretending the two types are a single type at the interface level, while happily separating the two in terms of (internally) expected behavior at the implementation level.

**Comment by brianharvey** [comment]
No, the behavior that's visible to users is the same.  In particular, setting a variable to a list value doesn't copy the list regardless of whether it's internally a dynamic array or a linked list, which is the thing you were surprised by in the first place.

The thing is, we're not trying to be a C/C++/Java/etc.  We're a Lisp.  Variable assignment is always a constant-time operation.  And our users are beginning CS students, so they haven't learned all those bad C/C++/Java/etc ways of thinking!

Other than efficiency, the only possibility of user-visible differences involves mutation.  So far, at least, we haven't implemented set-car! and set-cdr!, so if you want to do mutation you want your list in array form.  But that's just what you'll get if you do imperative programming.  What we tell users is, for any given list, decide if you're doing functional programming or imperative programming, and stick to it.

**Comment by micahcowan** [comment]
> In particular, setting a variable to a list value doesn't copy the list regardless of whether it's internally a dynamic array or a linked list, which is the thing you were surprised by in the first place.

Didn't surprise me. Maybe you mean the OP?

> The thing is, we're not trying to be a C/C++/Java/etc. We're a Lisp. 

Very aware of that (and heartily approve). But our lists are effectively both at the same time, which is what I was referring to: in other languages (including Lisps), separate list behaviors (like array vs linked) are in separate types. So long as our lists behave identically in all cases (ignoring performance) , I have no complaint. It just sounded to me like you were saying there's an actual behavior difference (which, in fact, there does seem to be).

>What we tell users is, for any given list, decide if you're doing functional programming or imperative programming, and stick to it.

Great. But as I said before, what if they're using library blocks they didn't write? Or are themselves writing functions to be used as libraries?

I mean, I suppose in theory,  you know whether a library function is meant to mutate your args or not, so you should still behave accordingly... I just feel like different types should be different... types.

Separating them this late in the game is obviously not realistic... but maybe if `List.prototype.cdr` were to convert the original list to linked, in-place, rather than just returning a copy when it's not linked... I mean, it's always going to be only one more item to copy, and you're doing the copy either way, the difference is that in _all_ cases, mutating the results of an `all but first of L` will also mutate `L`... instead of just sometimes, in a fashion that can be difficult to predict (particularly if your code uses library blocks written by others).

Thus unifying the unnecessarily divergent behavior (ignoring the necessary differences in performance, that are already present)...?

**Comment by brianharvey** [comment]
Right, I did mean the OP, sorry, I forgot that wasn't you.  (Too long a page to scroll up. :-))

Oh, I think maybe I didn't express myself clearly enough.  When I said it copies, I mean it does an O(n) time traversal of the data to convert its format.  Pointers to the list still work after the copying.

OTOH, I guess it's problematic if you make a pointer to the cdr and _then_ do something that converts to array form.  After that, the two versions are separate.  You could make a case that that's the right thing, because if you're programming functionally (which is what would lead you to cdr the list) you don't expect anything to change behind your back.  If you're programming functionally then you don't care whether the list pieces are eq; you only care about their values.

As I said, if we ever invent mutation for linked lists, this will require rethinking.

**Comment by micahcowan** [comment]
Yeah, that's what I'm getting at, sorry: if you cdr a linked list, and then mutate that, the conversion to array of the cdr is performed in-place, meaning that the original linked list is also changed (if I read the code correctly). But if you cdr an array, you receive a newly-`cons`ed list that is entirely a copy of (all but the first of) the original, and so the original is not modified. I could see how that behavior could be surprising, and it may not always be entirely under a single user/author's control as to whether this will happen.

(Note that the argument that the current behavior is "the right thing" would probably work much better if the behavior were reversed: if cdr-then-mutate on arrays changed the original and left original linked lists unmutated)

At this point, perhaps it's better if I file a new issue/pull request so we have some concrete code proposals to discuss. I feel bad for hijacking this thread and filling it with CS jargon 😄

**Comment by brianharvey** [comment]
Sigh.  Thank you for keeping pushing on this.  Just to be specific, I tried these:
![array-mod](https://cloud.githubusercontent.com/assets/3885074/22387010/7c3b1adc-e48e-11e6-8691-91fc7bb66d71.png)
![linked-mod](https://cloud.githubusercontent.com/assets/3885074/22387015/844c3102-e48e-11e6-9ef0-f8360c3f25b9.png)
<img width="291" alt="watchers" src="https://cloud.githubusercontent.com/assets/3885074/22387020/8e8dad8a-e48e-11e6-9a68-4ac12af847bf.png">
Sure enough, the value of the original list depends on whether it was linked before taking its cdr.

I suppose one way to fix this would be to change foo itself to linked when cdring it, so it would share memory with baz.  Since arrays never share memory, I don't _think_ that would break anything.

By the way, there's another problem, on which I'll start a new bug report:  If you pull off the last block of the second script, run the rest of it, and then separately run the replace, garply's watcher is not updated.

**Comment by micahcowan** [comment]
Thanks @brianharvey - yes, that's exactly what I was getting at.

> I suppose one way to fix this would be to change foo itself to linked when cdring it, so it would share memory with baz. Since arrays never share memory, I don't think that would break anything.

Yup, that's what I was thinking. There's no way to get a "deep" cdr without having got the cdr of the top of the array originally, so I think we should always be able to reconvert the entire array to cons cells.

The only conceivable case where that wouldn't be true, would be persisted state from a project written on current _Snap!_ versions, in which case the cdr-copy thing would already have taken place. I don't think that's a serious concern, though - presumably, any such project's author is already happy with their existing state, or if they're not, can regenerate it.

**Comment by micahcowan** [comment]
(The first pull request mentioned above is a misfire. Apparently an automated tool decided it'd be cool to muck with my global git config. Grrr....)

**Comment by brianharvey** [comment]
Thanks.  Jens is reluctant to change anything right now but we'll see about it later.

**Comment by brianharvey** [comment]
On reflection:  In Scheme it is officially an error to mutate a quoted list.  That is, you can't say (set! foo '(a b c d)) and later (set-car! (cdr foo) 'baz).  (The reason is that implementations may choose, or may not choose, to detect two uses of '(a b c d) and allocate the list only once.)  This situation is analogous to that, so we aren't required to behave in any particular way.  So we're going to postpone thinking about this for a while.
